corehq/apps/users/user_data.py

```diff
@@ -60,19 +67,45 @@ def save(self):
 
     @property
     def _provided_by_system(self):
-        return {
+        provided_data = {
             **(self.profile.fields if self.profile else {}),
             PROFILE_SLUG: self.profile_id or '',
             COMMCARE_PROJECT: self.domain,
         }
 
+        def _add_location_data():
+            if self._couch_user.get_location_id(self.domain):
+                provided_data[COMMCARE_LOCATION_ID] = self._couch_user.get_location_id(self.domain)
+                provided_data[COMMCARE_PRIMARY_CASE_SHARING_ID] = self._couch_user.get_location_id(self.domain)
+
+            if self._couch_user.get_location_ids(self.domain):
+                provided_data[COMMCARE_LOCATION_IDS] = user_location_data(
+                    self._couch_user.get_location_ids(self.domain))
+
+        # Some test don't have an actual user existed
+        # Web User don't store location fields in user data
+        if (self._couch_user or not settings.UNIT_TESTING) and self._couch_user.is_commcare_user():
+            _add_location_data()
+
+        return provided_data
+
+    @property
+    def _system_keys(self):
+        return set(self._provided_by_system.keys()) | LOCATION_KEYS
+
     def to_dict(self):
-        return {
+        combined_dict = {
             **self._schema_defaults,
             **self._local_to_user,
             **self._provided_by_system,
         }
 
+        # Ensure location keys are only from _provided_by_system
+        for key in LOCATION_KEYS:
+            if key in combined_dict and key not in self._provided_by_system:
+                del combined_dict[key]
```

Comment:
The idea is that system fields shouldn't come from `_local_to_user`, so it'd make sense to me to remove them only from there, rather than the combined dict, something like this:
```python
**{k: v for k, v in local_to_user.items() if k not in self._system_keys}
```
You could do that here, or even in `__init__`, which would also mean these would be removed from other places, like `raw` and the places that modify `_local_to_user`.  It'd also mean that these fields would be removed from the DB whenever user data models are saved (which might be helpful with the migration).

As-is, I think doing `del user_data['commcare_location_id']` won't raise an error, though it probably should.

Reply:
Hi Ethan, I love you advice and implement in bdba44ea47df35f2fde857e43d40ce325d3b66fa. However I didn't add it in __init__, I think it is an overkill, as we stop writing location info into UserData in hq, and also forbid this behavior in `__set__`, `update`.